server/util/bazel_request/bazel_request.go

```diff
@@ -20,25 +21,55 @@ var (
 
 const RequestMetadataKey = "build.bazel.remote.execution.v2.requestmetadata-bin"
 
+func getRequestMetadataBytes(ctx context.Context) []byte {
+	vals := metadata.ValueFromIncomingContext(ctx, RequestMetadataKey)
+	if len(vals) == 0 {
+		return nil
+	}
+	return []byte(vals[0])
+}
+
 func GetRequestMetadata(ctx context.Context) *repb.RequestMetadata {
-	if grpcMD, ok := metadata.FromIncomingContext(ctx); ok {
-		rmdVals := grpcMD[RequestMetadataKey]
-		for _, rmdVal := range rmdVals {
-			rmd := &repb.RequestMetadata{}
-			if err := proto.Unmarshal([]byte(rmdVal), rmd); err == nil {
-				return rmd
-			}
-		}
+	b := getRequestMetadataBytes(ctx)
+	if len(b) == 0 {
+		return nil
+	}
+	rmd := &repb.RequestMetadata{}
+	if err := proto.Unmarshal(b, rmd); err != nil {
+		return nil
 	}
-	return nil
+	return rmd
 }
 
 func GetInvocationID(ctx context.Context) string {
```

Comment:
What if we pre-extracted the invocation ID in one of our interceptors and set it as a value on the context -- then this method could check for the presence of that value on the context before extracting grpc MD and unmarshaling?

Seems like a simpler solution than this proto stuff (all though I am into this lol)

Reply:
yep, exactly

Reply:
+1 to your suggestion re internal RPCs